core/src/main/java/org/apache/iceberg/metrics/MetricsReporters.java

```diff
@@ -0,0 +1,101 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.metrics;
+
+import java.util.Set;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.immutables.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Utility class that allows combining two given {@link MetricsReporter} instances. */
+public class MetricsReporters {
+  private MetricsReporters() {}
+
+  public static MetricsReporter combine(MetricsReporter first, MetricsReporter second) {
+    if (null == first) {
+      return second;
+    } else if (null == second || first == second) {
+      return first;
+    }
+
+    if (first instanceof CompositeMetricsReporter && second instanceof CompositeMetricsReporter) {
+      return combineComposites((CompositeMetricsReporter) first, (CompositeMetricsReporter) second);
+    }
+
+    if (first instanceof CompositeMetricsReporter) {
+      return combineWithSimpleReporter((CompositeMetricsReporter) first, second);
+    }
+
+    if (second instanceof CompositeMetricsReporter) {
+      return combineWithSimpleReporter((CompositeMetricsReporter) second, first);
+    }
+
+    return combineSimpleReporters(first, second);
+  }
+
+  private static CompositeMetricsReporter combineSimpleReporters(
+      MetricsReporter first, MetricsReporter second) {
+    Set<MetricsReporter> result = Sets.newIdentityHashSet();
+    result.add(first);
+    result.add(second);
+
+    return ImmutableCompositeMetricsReporter.builder().metricsReporters(result).build();
+  }
+
+  private static CompositeMetricsReporter combineComposites(
+      CompositeMetricsReporter first, CompositeMetricsReporter second) {
+    Set<MetricsReporter> result = Sets.newIdentityHashSet();
+    result.addAll(first.metricsReporters());
+    result.addAll(second.metricsReporters());
+
+    return ImmutableCompositeMetricsReporter.builder().metricsReporters(result).build();
+  }
+
+  private static CompositeMetricsReporter combineWithSimpleReporter(
+      CompositeMetricsReporter first, MetricsReporter second) {
+    Set<MetricsReporter> result = Sets.newIdentityHashSet();
+    result.addAll(first.metricsReporters());
+    result.add(second);
+
+    return ImmutableCompositeMetricsReporter.builder().metricsReporters(result).build();
+  }
+
+  @Value.Immutable
+  public interface CompositeMetricsReporter extends MetricsReporter {
+    Logger LOG = LoggerFactory.getLogger(CompositeMetricsReporter.class);
+
+    Set<MetricsReporter> metricsReporters();
```

Comment:
What is the value of using immutables for this?

The alternative is this:

```java
  private static class CompositeMetricsReporter extends MetricsReporter {
    private final Set<MetricsReporter> reporters;

    private CompositeMetricsReporter(Set<MetricsReporter> reporters) {
      this.reporters = reporters;
    }

    @Override
    public void report(MetricsReport report) { ... }
  }
```

I don't think there is any value in immutables, but there is a cost because this now has autogenerated code that you can't actually see and the uses must all go through a builder that's also generated and not shown.

Reply:
in this case it's probably indeed better to not use an Immutable class since instantiation is guarded by the factory method